### PR TITLE
API: Implement sorting

### PR DESCRIPTION
**Pull request by seberg**
This implements sorting based on the same approach as cupynumeric.  There are a few smaller rough edges, but it is getting there.

Things that are still needed:
* Support of column and sort order (internals understand it).
* Some smaller cleanups maybe (e.g. `shuffle` is used from `repartition_by_hash.cu`).
* We need to broadcast splits to all, this abuses `shuffle` (not sure this is a big deal).
* Some simple C-tests are needed
* Python tests are good about checking the splits (i.e. the window).  But need to add randomized data and tests for the missing API (column/null order).

### CODE REVIEW

File: cpp/src/sort.cpp [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+ * @param include_start Whether to include the starting 0.
+ * @returns cudf column selecting containing nsplits indices.
+ */
+std::unique_ptr<cudf::column> get_split_ind(GPUTaskContext& ctx,
+                                            cudf::size_type nvalues,
+                                            int nsplits,
+                                            bool include_start)
+{
+  auto nvalues_per_split = nvalues / nsplits;
+  auto nvalues_left      = nvalues - nvalues_per_split * ctx.nranks;
```
[the remark this answers is not in the document]
Comment (seberg) [reply cpp/src/sort.cpp:64]: Yeah, must always be the same, but for the helper method I think it is nicer to pass it (and not use `ctx.nranks`).

File: cpp/src/sort.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+                                        ctx.mr());
+
+  auto my_split_cols_view = my_split_cols_tbl->view();
+  auto my_split_cols_vector =
+    std::vector<cudf::column_view>(my_split_cols_view.begin(), my_split_cols_view.end());
+
+  // Add in rank and local index (together provide a global order).
+  my_split_cols_vector.push_back(my_split_rank_col->view());
+  my_split_cols_vector.push_back(my_split_ind_col->view());
+  auto my_splits = cudf::table_view(my_split_cols_vector);
```
[the remark this answers is not in the document]
Comment (seberg) [reply cpp/src/sort.cpp:160]: Right, I thought that was right because it leads to even splits for already sorted data:
`[1, 2, 3, 4], [5, 6, 7, 8]` -> split candidates `[1, 2, 3, 4] + [5, 6, 7, 8]` -> split point is chosen to be at (before) `5`.

If I would not include it, the split point chosen there would be `6` meaning data is moved?
When the data trivially the same `[1, 2, 3, 4], [1, 2, 3, 4]` -> `[1a, 1b, 2a, 2b, 3a, 3b, 4a, 4b]`
 -> Split before `3a`, which gives me an even result with `[1, 1, 2, 2], [3, 3, 4, 4]` as well?

File: cpp/src/sort.cpp [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+  auto split_candidates_last_view  = split_candidates_last_col->view();
+
+  // The local index and rank of the split value, we'll use the rank if it came from this rank
+  auto split_candidates_equal_view = split_values_view.column(my_splits.num_columns() - 1);
+  auto split_candiates_rank_view   = split_values_view.column(my_splits.num_columns() - 2);
+
+  /*
+   * Copy all the above information to the host and finalize the local splits.
+   */
+  auto nsplitpoints = ctx.nranks - 1;
```
[the remark this answers is not in the document]
Comment (seberg) [reply cpp/src/sort.cpp:234]: Yeah, that is right:  `cudf:split` returns `N+1` partitions, so no need to include the start or end.

File: cpp/src/sort.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (seberg) [reply cpp/src/sort.cpp:160]: OK, so I was indeed doing `<` splits excluding the last point (the loop is `nsplits-1`).  So with subsampling, yeah, I get `[0, 3] and [5, 7] -> [5] + end`.
Since `libcudf` slicing is exclusive (and implicitly includes the end slice) that is the clearer way around here.

So, I think this is all the same, then?  Since I use `<` split points, I just include the first here for the candidates, rather than the last that you use with `<=` splits.

**Comment by seberg** [comment]
@madsbk if you have time at some point, would be nice if you can have a look.  I should duplicate the shuffling to have a proper broadcast-shuffle, but I think that is the only issue.

I discussed correctness with @mfoerste4, so I think review is just about code and not logic (like how we split exactly).

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -71,7 +71,7 @@ include(cmake/Modules/ConfigureCUDA.cmake) # set other CUDA compilation flags
 # * dependencies ----------------------------------------------------------------------------------
 
 rapids_find_package(
-  legate REQUIRED Legion LegionRuntime
+  legate
```
Comment (madsbk) [inline cpp/CMakeLists.txt:74]:
```suggestion
  legate REQUIRED
```

Comment (madsbk) [reply cpp/CMakeLists.txt:74]: Isn't Legate still required?

File: cpp/include/legate_dataframe/sort.hpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <legate_dataframe/core/table.hpp>
+
+namespace legate::dataframe {
+
+LogicalTable sort(const LogicalTable& tbl,
+                  const std::vector<std::string>& keys,
+                  const std::vector<cudf::order>& column_order,
+                  const std::vector<cudf::null_order>& null_precedence,
+                  bool stable = false);
```
Comment (madsbk) [inline cpp/include/legate_dataframe/sort.hpp:30]: It would be good with a docstring of `sort`

**Review by madsbk — CHANGES_REQUESTED** [review]
Nice work! I only have some minor suggestions

File: cpp/CMakeLists.txt (hunk 4 above)
Comment (seberg) [reply cpp/CMakeLists.txt:74]: I'll revert this.  I am honestly not sure whether it is, but this was an (unsuccessful) try to fix the linking issue with current legate.

**Review by madsbk — APPROVED** [review]

**Comment by seberg** [comment]
I opened gh-19 for the task of cleaning up the shuffling a bit.  Maybe easier as a follow-up (but don't hesitate to kick me to do it or any other follow-ups here!).
I think we might re-organize the whole `shuffle` code soon anyway, though.
